packages/vats/src/lib-chainPublish.js

```diff
@@ -0,0 +1,46 @@
+// @ts-check
+
+import { E } from '@endo/far';
+
+/**
+ * Publish results of an async iterable into a chain storage node as an array
+ * serialized using JSON.stringify or an optionally provided function (e.g.,
+ * leveraging a serialize function from makeMarshal).
+ *
+ * @param {AsyncIterator} iterable
+ * @param {ReturnType<typeof import('./lib-chainStorage.js').makeChainStorageRoot>} chainStorageNode
+ * @param {{ timerService: ERef<TimerService>, serialize?: (obj: any) => string }} powers
+ */
+export async function publishToChainNode(
```

Comment:
This was originally envisioned as a function that accepts a chainStorage node and returns a subscription kit, but upon digging in it seemed more useful to me for it to consume from an existing iterable and return nothing. Still another option would be to make it effectively a tee, but I think it makes more sense to have the caller compose e.g. `getKey()` and the base iterable.

Reply:
What layers do you think are being crossed and what concerns are being combined? Transmitting arbitrary data into a chainStorage node seems very respectful of layers and independent concerns. As for #5400, I looked it over carefully but didn't see anything to address the specific requirements of publishing to chain storage... am I just missing something?

Reply:
> What layers do you think are being crossed and what concerns are being combined?

I guess what jumps out at me looking at the helper is it is trying to combine the iterator observation with the block batching.  I think that the block batching belongs in the cosmic-swingset layer, and the iteration client is handled by #5400.

> Transmitting arbitrary data into a chainStorage node seems very respectful of layers and independent concerns.

You're right that it needs to be done at a lower layer than #5400, but it should be even lower than JS entirely.  I'm trying hard to "put the code where the data is".  Since it is a cosmic-swingset concern (touching block height, event log, and merkle tree), it belongs outside of vats.  I appreciate you're trying to respect peoples' time, but I would really like it if you would reach out if something about the layering doesn't make sense or seems unnecessarily weird (e.g. the chainTimer hack).

> As for #5400, I looked it over carefully but didn't see anything to address the specific requirements of publishing to chain storage... am I just missing something?

Here is an adapter that explains how #5400 accomplishes these platform concerns, presuming that the batching is solved at the cosmic-swingset layer:

```js
// adapt chainStorageNode to makeMarshalSubscriber
const publisher = Far('chainPublisherNode', {
  publish: jsonable => E(chainStorageNode).setValue(JSON.stringify(jsonable)),
  getPublisherDetails: () => { chainStorageKey: E(chainStorageNode).getKey() },
});

// Then something like:
makeMarshalSubscriber(observeEach(subscription), publisher);
```

Is that clearer now?  I'll be working on splitting `x/pubstore` Golang module from `x/swingset`.  I want to implement cosmos-specific publishing concerns there, such as batching, history, and event logging.

With some polish and moving the code around until we have minimal interfaces, I think this could be quite good.